### PR TITLE
Test slider

### LINKED ISSUE #262: Tests for Slider Widget

### PR DESCRIPTION
**Pull request by hmbui**
Also close #262 .

### CODE REVIEW

File: pydm/widgets/slider.py [hunk 1]
```diff
@@ -240,7 +245,7 @@ def alarm_severity_changed(self, new_alarm_severity):
         """
         PyDMWritableWidget.alarm_severity_changed(self, new_alarm_severity)
         if hasattr(self, "value_label"):
-            if self._channels is not None:
+            if self._channel is not None:
```
Comment (hhslepicka) [inline pydm/widgets/slider.py]: Please add a test for an empty string also.

File: pydm/widgets/slider.py [hunk 2]
```diff
@@ -457,7 +462,7 @@ def userMinimum(self, new_min):
         ----------
         new_min : float
         """
-        self._user_minimum = float(new_min)
+        self._user_minimum = float(new_min) if new_min is not None else None
```
Comment (hhslepicka) [inline pydm/widgets/slider.py]: What is the consequence of None in here at `self._user_minimum`?

File: pydm/widgets/slider.py [hunk 3]
```diff
@@ -481,7 +486,7 @@ def userMaximum(self, new_max):
         ----------
         new_max : float
         """
-        self._user_maximum = float(new_max)
+        self._user_maximum = float(new_max) if new_max is not None else None
```
Comment (hhslepicka) [inline pydm/widgets/slider.py]: What is the consequence of None in here at `self._user_maximum`?

File: pydm/tests/widgets/test_slider.py [hunk 4]
```diff
@@ -0,0 +1,659 @@
...
+    signals.new_value_signal.connect(pydm_slider.channelValueChanged)
+    signals.new_value_signal.emit(new_channel_value)
+
+    assert pydm_slider.value_label.text() == pydm_slider.format_string.format(pydm_slider.value)
+    if not is_slider_down:
+        expected_slider_value = np.argmin(abs(pydm_slider._slider_position_to_value_map - float(new_channel_value)))
+        assert pydm_slider._slider.value() == expected_slider_value
+    else:
+        assert pydm_slider._slider.value() == 0
+
```
Comment (mattgibbs) [inline pydm/tests/widgets/test_slider.py]: A potential issue I always worry about with the slider is that the channel's value will change via some external means, which causes the slider to update, and somehow the slider's position change causes the widget to change the value again.  It would be nice to hook up a slot to the slider's `send_value_signal` that writes to `new_channel_value`, and at the end of the test, ensure that `new_channel_value` has not changed.

File: pydm/widgets/slider.py (hunk 3 above)
Comment (mattgibbs) [reply pydm/widgets/slider.py]: In `reset_slider_limits`, if the minimum or maximum is none (from the user, or from the channel), it puts the widget into a 'needs limit info' state, which disables the widget.  The internal slider's minimum or maximum value just remain wherever they were (likely the default values) until the min/max gets set to a non-None value.

So, setting this to None should be fine.

File: pydm/widgets/slider.py (hunk 2 above)
Comment (mattgibbs) [reply pydm/widgets/slider.py]: See my comment for self._user_maximum.

File: pydm/widgets/slider.py (hunk 3 above)
Comment (hhslepicka) [reply pydm/widgets/slider.py]: Awesome! Thank you for clarifying, @mattgibbs. 👍

File: pydm/tests/widgets/test_slider.py (hunk 4 above)
Comment (hmbui) [reply pydm/tests/widgets/test_slider.py]: I think I've already addressed this concern with the test_internal_slider_value_changed() test. @mattgibbs, could you take another look?

**Comment by hmbui** [comment]
I've gone through your comments. As the user min and max can be None, and that I also have already created a test where the slider can move while updating the channel value, I don't think there's anything I should do here. @hhslepicka please let me know if I'm missing anything. Thanks.

**Comment by hhslepicka** [comment]
@hmbui what about this comment here https://github.com/slaclab/pydm/pull/327#discussion_r191844561 ?

**Comment by hmbui** [comment]
Actually, I forgot -- I had added the test cases when the channel address is an empty string.

So, I've just pushed the new test cases in response to @hhslepicka's comment for that case.

**Comment by hmbui** [comment]
@mattgibbs @hhslepicka @teddyrendahl Hi guys, I've addressed your change requests. Please take another look. Thanks!
